static/js/narrow.js

```diff
@@ -234,11 +249,18 @@ exports.activate = function (raw_operators, opts) {
         ui.show_loading_more_messages_indicator();
     }
 
-    // Put the narrow operators in the URL fragment.
-    // Disabled when the URL fragment was the source
-    // of this narrow.
+    // Put the narrow operators in the URL fragment. Disabled when the URL
+    // fragment was the initiating source of this narrow.
     if (opts.change_hash) {
         hashchange.save_narrow(operators);
+    } else if (following_alias) {
+        // We want this to update the hash on page load if someone followed
+        // a link to a renamed stream. Since hashchange will refuse us right now
+        // (it's already tracking a page load), we process on a future tick.
+        // We set the "replace_state" parameter true to avoid polluting history.
+        setTimeout(function () {
+            hashchange.save_narrow(operators, true);
```

Comment:
This loads the new narrow directly (the Filter has been intercepted and updated by the time we reach this line, in the first call to `narrow.activate`). It's just the hash that updates in the next tick; the data are already loaded. In testing this, I don't see any flicker.

I've looked at moving this logic to onload (have this implemented locally; can share), but some concerns:

 * hashchange.js is indeed thorny. I don't love adding more logic to the `changing_hash` detection
 * moving logic to onload means that typing an old stream name into the narrow won't forward it; this leads to an inconsistency where Zulip displays an empty when a user types an old stream name into the search box (e.g. "stream:OldName"), but the upon hitting Ctrl-R the forward happens, the URL updates, and content appears. 

Thoughts @timabbott ?